plugins/src/main/java/opengrok/auth/plugin/ldap/LdapServer.java

```diff
@@ -122,13 +129,89 @@ public void setInterval(int interval) {
         this.interval = interval;
     }
 
+    private String urlToHostname(String urlStr) throws URISyntaxException {
+        URI uri = new URI(urlStr);
+        return uri.getHost();
+    }
+
+    /**
+     * This method converts the scheme from URI to port number.
+     * It is limited to the ldap/ldaps schemes.
+     * The method could be static however then it cannot be easily mocked in testing.
+     * @return port number or -1 if the scheme in given URI is not known
+     * @throws URISyntaxException if the URI is not valid
+     */
+    public int getPort() throws URISyntaxException {
+        URI uri = new URI(getUrl());
+        switch (uri.getScheme()) {
+            case "ldaps":
+                return 636;
+            case "ldap":
+                return 389;
+        }
+
+        return -1;
+    }
+
+    private boolean isReachable(InetAddress addr, int port, int timeOutMillis) {
+        try {
+            try (Socket soc = new Socket()) {
+                soc.connect(new InetSocketAddress(addr, port), timeOutMillis);
+            }
+            return true;
+        } catch (IOException e) {
+            return false;
+        }
+    }
+
+    /**
+     * Wraps InetAddress.getAllByName() so that it can be mocked in testing.
+     * (mocking static methods is not really possible with Mockito)
+     * @param hostname hostname string
+     * @return array of InetAddress objects
+     * @throws UnknownHostException if the host cannot be resolved to any IP address
+     */
+    public InetAddress[] getAddresses(String hostname) throws UnknownHostException {
+        return InetAddress.getAllByName(hostname);
+    }
+
+    /**
+     * Go through all IP addresses and find out if they are reachable.
+     * @return true if all IP addresses are reachable, false otherwise
+     */
+    public boolean isReachable() {
+        try {
+            for (InetAddress addr : getAddresses(urlToHostname(getUrl()))) {
+                // InetAddr.isReachable() is not sufficient as it can only check ICMP and TCP echo.
+                int port = getPort();
+                if (!isReachable(addr, port, getConnectTimeout())) {
+                    LOGGER.log(Level.WARNING, "LDAP server {0} is not reachable on {1}:{2}",
+                            new Object[]{this, addr, Integer.toString(port)});
+                    return false;
+                }
+            }
+        } catch (UnknownHostException e) {
+            LOGGER.log(Level.SEVERE, String.format("cannot get IP addresses for LDAP server %s", this), e);
+            return false;
+        } catch (URISyntaxException e) {
+            LOGGER.log(Level.SEVERE, String.format("not a valid URI: %s", getUrl()), e);
+            return false;
+        }
+
+        return true;
```

Comment:
Could `getAddresses()` be empty, and would `true` be correct then?

Reply:
I noticed this too. Thanks for reminding me.